conjure-java-core/src/test/java/com/palantir/conjure/java/types/Union2.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.types;
+
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonAnySetter;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonSetter;
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+import com.palantir.conjure.java.types.Union2.Bar;
+import com.palantir.conjure.java.types.Union2.Baz;
+import com.palantir.conjure.java.types.Union2.Foo;
+import com.palantir.conjure.java.types.Union2.UnknownVariant;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.Safe;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nonnull;
+
+/**
+ * This is hand-rolled, I just want to make sure it's compatible.
+ */
+@JsonTypeInfo(
+        use = JsonTypeInfo.Id.NAME,
+        include = JsonTypeInfo.As.EXISTING_PROPERTY,
+        property = "type",
+        visible = true,
+        defaultImpl = UnknownVariant.class)
+@JsonSubTypes({@JsonSubTypes.Type(Foo.class), @JsonSubTypes.Type(Bar.class), @JsonSubTypes.Type(Baz.class)})
+@JsonIgnoreProperties(ignoreUnknown = true)
+public sealed interface Union2 {
+
+    sealed interface Known permits Foo, Bar, Baz {}
```

Comment:
```suggestion
    sealed interface Known extends Union2 permits Foo, Bar, Baz {}
```

Then below we don't need to implement `Union2, Known`, only `Known`

Reply:
Idea here is that we can use `MyUnion.Known` internally for business logic, ensuring that nothing has to validate against unknown variants, but no casting/validation is required to return or pass `MyUnion.Known` as a `MyUnion`.

We may be able to do something clever with safety annotations on `MyUnion.Known` that can't be guaranteed on the base type with an unknown variant (some issues today since we don't allow safety to be redefined as safer than a supertype, but I think it's reasonable in this instance, trick is ensuring we know when is safe, and when isn't)